### PR TITLE
Support for adding accessibility features at chart generation

### PR DESCRIPTION
**Pull request by gherka**
Currently, to add accessibility features to dc.js charts you need to do it as part of post-processing. It would be easier to simply specify what features you'd like to include together with other chart attributes, like width and height and labels. This idea has been brought up before (#1185) and after raising it again on the mailing list recently, @gordonwoodhull encouraged me to try to implement something.

I'm thinking of a series of commits / PRs to cover the basics:

- This one makes all SVG charts accessible by keyboard using tabindex. It also adds a description that screen readers can access. The configuration of a `<desc>` element and `aria-labelledby` attribute seems to have the best support according to [this](https://www.deque.com/blog/creating-accessible-svgs/). I tried adding `<title>`, but it fails a few tests looking at tooltips and besides, some screen readers will read it twice. It defaults to an empty string and ideally, I'd like it to reflect the default chart type - so a `dc.PieChart` will have "pie chart" in the description and so on. This will likely require modifying each chart file with some default attribute, however. Users can add a description for their charts by simply passing a string to `.svgDescription()`.

- Make internal chart elements, like `<circle>` in a bubble chart or `rect` in a bar chart keyboard-selectable. Also, make them "clickable" from keyboard.

- When I tested tooltips created using `<title>` in Narrator / NVDA, I got mixed results for a variety of browsers. In some, the tooltip text was read out, on others it wasn't. There is probably a way to make it a little bit more consistent.

- Optionally designating the `NumberDisplay` widget as `aria-live` [region](https://developer.paciellogroup.com/blog/2014/03/screen-reader-support-aria-live-regions/), meaning changes to it will be read out by screen readers.

Here's the summary of this PR as it currently stands:

- [x] root SVG can get focus and users can add custom description using `svgDescription` method

- [x] internal chart elements can get focus and filtering interactions will trigger on `Enter` or `Space` as if clicked

- [x] CSS will behave on focus in the same way as it behaves on hover

- [x] Individual chart types are accessible  

  - [x] Bar chart
  - [x] Box plot
  - [x] Bubble chart
    - [x] DOM order of bubbles can be changed based on x value
  - [x] Choropleth
  - [x] Heatmap
  - [x] Line chart
  - [x] Pie chart
  - [x] Row chart
  - [x] Scatter plot
    - [x] DOM order of bubbles can be changed based on x value
  - [x] Sunburst chart
 <html><br></html>

- [x] NumberDisplay can optionally be set as `aria-live` region

- [x] Legends can get focus and interactivity

- [x] Add example

### CODE REVIEW

**Comment by gordonwoodhull** [comment]
Haven’t tested yet, but it looks good!

I think you can default `svgDescription` to something reasonable with `this.constructor.name`

https://stackoverflow.com/a/10314492

**Comment by gherka** [comment]
Good idea! I've added it and set a basic test up on a bar chart. I also checked other chart examples (pie, line, row, composite) and it works as expected.

**Comment by gherka** [comment]
To make significant internal chart elements like bars in a bar chart accessible, I've added a dc prefixed class in the chart's module, but since the effects of this feature are the same across all chart types (`tabindex=0`, triggers `onClick` on keypress), the function is part of `base-mixin.js`. This way, if a new chart type comes along, its creator would only need to think about which elements are "significant" and simply add the `dc-tabbable` class to them. 

Also, elements created at the initial render can disappear and new ones appear after filtering - so the new function would need to run on both `render()` and `redraw()`. 

@gordonwoodhull, if you approve of this approach, I'll go through all existing chart types and amend them in the same way as I did with the bar chart, and write some tests.

File: src/base/base-mixin.js [hunk 1]
```diff
@@ -1,4 +1,4 @@
-import {select} from 'd3-selection';
+import {select, event as d3Event} from 'd3-selection';
```
Comment (gherka) [inline src/base/base-mixin.js:1]: Had to alias "event" because it's used as a function parameter elsewhere.

**Comment by gordonwoodhull** [comment]
I hope to take a look at this today, but I’m out sick, so please bear with me if it takes a little longer.

**Comment by gherka** [comment]
Of course, no worries at all. Hope you get better soon!

File: src/base/base-mixin.js [hunk 2]
```diff
@@ -521,10 +523,45 @@ export class BaseMixin {
 
     generateSvg () {
         this._svg = this.root().append('svg');
+
+        this._svg.append('desc')
+            .attr('id', `desc-id-${this.__dcFlag__}`)
+            .html(`${this._svgDescription}`);
+
+        this._svg
+            .attr('tabindex', '0')
```
Comment (gordonwoodhull) [inline src/base/base-mixin.js:532]: Should this also be enabled/disabled by `keyboardAccessible`?

My concern is that it will probably change the behavior of existing applications.

File: src/base/base-mixin.js [hunk 3]
```diff
@@ -663,11 +700,34 @@ export class BaseMixin {
...
         return result;
     }
 
+    _makeKeyboardAccessible () {
+        this._svg.selectAll('.dc-tabbable')
+                .attr('tabindex', 0)
+                .on('keydown', d => {
+
+                    if (d3Event.keyCode === 13) {
+                        console.log('clicked!');
```
Comment (gordonwoodhull) [inline src/base/base-mixin.js:718]: did you mean to check in the debug logs?

File: src/base/base-mixin.js [hunk 4]
```diff
@@ -521,10 +523,45 @@ export class BaseMixin {
...
+        this._svg
+            .attr('tabindex', '0')
+            .attr('role', 'img')
+            .attr('aria-labelledby', `desc-id-${this.__dcFlag__}`);
         this.sizeSvg();
         return this._svg;
     }
 
+    /**
+     * Set description text for the entire SVG graphic to be read out by assistive technologies.
```
Comment (gordonwoodhull) [inline src/base/base-mixin.js:540]: Please document the default here (in prose)

File: src/base/base-mixin.js (hunk 2 above)
Comment (gherka) [reply src/base/base-mixin.js:532]: Sorry, my fault for not explaining the reasoning. I think it's useful to be able to turn the accessibility of internal chart elements on and off separately from the accessibility of the chart as a whole - depending on chart design, tabbing through many, many elements might not be desirable. `keyboardAccessible` is the method for the former and rolls together focusing and clicking. `svgDescription` is for the latter and covers focus and description. Since description is only useful on focused elements, I included `tabindex` by default.

The only scenario I can think of when this would result in an unexpected change is if the application changed `tabindex` to a value other than 0 which isn't [recommended](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/tabindex). I think as is, it's a "free" improvement for existing dashboards that mouse users shouldnt even notice.

Comment (gordonwoodhull) [reply src/base/base-mixin.js:532]: I think I get the design. What I meant is, what if someone has an existing dashboard with form elements, but the charts are not currently accessible using the tab key?

I think it's pretty common to have just a text entry field and a couple of selects, or whatever. I know I am lazy and I use the keyboard instead of the mouse in a familiar application, when I have to do something repetitive.

In other words, not all keyboard users are screen reader users.

With this change, the charts will be tabbable even though there isn't any interactivity or screen reader enabled for them. My concern is that it will make these simple form/dashboard applications harder to navigate, even though no functionality has been added.

Here's a very simple example: [complex reductions example](https://dc-js.github.io/dc.js/examples/complex-reduce.html). Before the change, just the links and the radio button group are accessible with the tab (on Chrome - on Firefox the links are not tabbable). After the change, the charts are tabbable but there isn't any functionality when they are focused.

Comment (gherka) [reply src/base/base-mixin.js]: Thanks for taking the time to explain. I've made a few changes (and tests) so that we don't add any new tabbable elements unless the user explicitly tells us to by either setting `svgDescription` or `keyboardAccessible`. The latter will keep the class constructor name as default description, unless it's overwritten by `svgDescription`. The implementation feels a bit hacky, though.

File: src/base/base-mixin.js [hunk 5]
```diff
@@ -521,10 +523,59 @@ export class BaseMixin {
...
+     * @returns {String|BaseMixin}
+     */
+    svgDescription () {
+
+        // overwrite for public API
+        this.svgDescription = description => {
+            this.svgDescription.userSet = true
+            this._svgDescription = description;
+            return this;
+        };
```
Comment (gordonwoodhull) [inline src/base/base-mixin.js:556]: This seems too complicated. When you call the method, it replaces itself with another method, that, when used as a setter, sets a flag on the function object? And then call it on initialization?

Why not do the following:

* Initialize `_svgDescription` to null
* When `.svgDescription()` is called as a getter, return `_svgDescription || this.constructor.name
* When called as a setter, assign to `_svgDescription` as usual
* When testing if it's been set, use `_svgDescription`, but when getting the value, call `.svgDescription()`

Comment (gherka) [reply src/base/base-mixin.js]: Well, this is embarrassing :) Was over-thinking it and saw a problem where there wasn't one, really.

**Comment by gordonwoodhull** [comment]
Looks great, thank you!

**Comment by gherka** [comment]
Had to modify the approach a little bit because the way `on('click', fn)` function is specified varies between concrete chart implementations - could be a straightforward `onClick(d)`, or `onClick(d, layerIndex)` or `_onClick(d)` or `onClick(d, i)` or just an anonymous function. Now in addition to adding a `dc-tabbable` class to the relevant elements, you also need to tell `_makeKeyboardAccessible` what to do on click.

A couple of other observations:
-  the DOM order of bubble chart circles is based on the key - what do you think of making it based on the x-value instead so that tab order goes predictably from left to right? We'd need to preserve the option of putting smaller-radius circles on top with `sortBubbleSize`
- since BubbleOverlay requires "external" SVG to function, `generateSVG` from BaseMixin isn't called so svgDescription isn't added, even if called. What's the best way to signpost this behaviour?

**Comment by gordonwoodhull** [comment]
Hi @gherka, sorry for the slow reply.

Yes, I think that ordering the bubble chart circles by X value would be okay, as long as `sortBubbleSize` is still respected. I don't think it makes any difference except for Z and tab order.

That's an interesting question about BubbleOverlay. I don't think this chart is used very much, so it might not matter too much, but I suppose you could do the annotation on render (or `_initOverlayG`) instead? I know it's ugly to have a special case for this one chart, but that's just what is, a single chart that doesn't create its own `svg`.

**Comment by gherka** [comment]
I think this now covers everything in the initial scope of features. There's just a couple of things I'm not too sure about - 
- What tests do I need to add? Some, like ensuring that `NumberDisplay` doesn't have any transitions when `keyboardAccessible` is `true` is quite tricky and haven't been able to figure out how to do it yet.
- What examples do I need to add? While developing, I've tested each feature on the `web-src` examples so they can all potentially work.
- What about resolving conflicts? Do I need to do it in my PR?

File: src/base/base-mixin.js [hunk 6]
```diff
@@ -668,6 +715,28 @@ export class BaseMixin {
...
+    _makeKeyboardAccessible (onClickFunction, ...onClickArgs) {
+        // called from each chart module's render and redraw methods
+        const tabElements = this._svg
+            .selectAll('.dc-tabbable')
+            .attr('tabindex', 0);
+                
+        if (onClickFunction) {
+            tabElements.on('keydown', (d, i) => {
+                // trigger only if d is an object undestood by KeyAccessor()
+                if (d3Event.keyCode === 13 && typeof d === 'object') {
```
Comment (gordonwoodhull) [inline src/base/base-mixin.js:727]: Maybe test if the key accessor returns truthy value? Just to be careful.

File: src/base/base-mixin.js [hunk 7]
```diff
@@ -668,6 +715,28 @@ export class BaseMixin {
...
     }
 
+    _makeKeyboardAccessible (onClickFunction, ...onClickArgs) {
+        // called from each chart module's render and redraw methods
+        const tabElements = this._svg
+            .selectAll('.dc-tabbable')
+            .attr('tabindex', 0);
+                
+        if (onClickFunction) {
+            tabElements.on('keydown', (d, i) => {
```
Comment (gordonwoodhull) [inline src/base/base-mixin.js:725]: Since you started this PR, D3 was updated to 6.0 and the signature of event handlers [changed](https://github.com/d3/d3/blob/master/CHANGES.md#changes-in-d3-60) - now the signature is `(event, d)`

We built an adapter for dc@4.1 called `adaptHandler` that smooths over the difference, to move to d3@6 but still support d3@5. I would suggest using it here.

**Review by gordonwoodhull — CHANGES_REQUESTED** [review]
One small suggestion, and a bigger change to update to dc@4.1.

**Comment by gordonwoodhull** [comment]
Hi @gherka!

Sorry for the slow reply - I am on vacation this week and next.

* We don't generally test animated transitions, so I think it is okay to skip that. The specific tests that you made for edge cases are great! It would also be good to test the general case for each chart, i.e. that each element has been modified. This will help to prevent regressions in the future when other people work on the code.
* I think it would be sufficient to add one example with a few charts - for example, maybe adding accessibility to the [filtering example](https://dc-js.github.io/dc.js/examples/filtering.html) - so that people can try the feature out.
* Yes, please rebase your branch onto develop and then force-push the result to resolve conflicts. The PR will update automatically.

Unfortunately, there was a big change in dc@4.1 which affects your PR. As I mentioned above, the signature of event handlers changed in d3@6, and we are using an adapter to continue support for dc@5 until we decide to drop it. Please use this adapter and test your code with d3@5 and d3@6 to make sure it functions correctly. We are no longer testing d3@5, so I'll take your word on this.

Thanks for your careful work on this feature - people have asked for this for a long time, and it was something that worried me personally. It's good to see that accessibility can be added without much disruption to other functionality.

**Comment by gherka** [comment]
Hi @gordonwoodhull, thank you for your feedback! In this commit I've added automatic tests for each chart type that I had modified and also tested each chart type on the standard examples using v5 and v6. As you suggested, I've included `keyboardAccessible()` and `svgDescription()` in the `filtering.html` example. 

There are a few minor things that can be tweaked further, but the base functionality is now in place, I think.

File: src/base/base-mixin.js (hunk 6 above)
Comment (gherka) [reply src/base/base-mixin.js]: I tried testing for it (`this.keyAccessor()(d)`), but it failed for bar and pie charts and `this.keyAccessor()` is always truthy because it returns a function. Is there another way?

File: spec/bar-chart-spec.js [hunk 8]
```diff
@@ -1278,6 +1278,81 @@ describe('dc.BarChart', () => {
...
+            regionDimension.filterExact('West');
+            chart.redraw();
+            regionDimension.filterAll();
+            chart.redraw();
+                    
+            chart.selectAll('rect.bar').each(function () {
+                const bar = d3.select(this).node();
+                bar.dispatchEvent(event);
+                expect(clickHandlerSpy).toHaveBeenCalled();            
+            });
```
Comment (gordonwoodhull) [inline spec/bar-chart-spec.js:1350]:
```suggestion
            chart.selectAll('rect.bar').each(function (d) {
                this.dispatchEvent(event);
                expect(clickHandlerSpy).toHaveBeenCalledWith(d);
            });

```
1. It's better to check that the correct thing has been clicked, rather than just checking if the spy has been called. Also,  `toHaveBeenCalled` will always be true after the first iteration, because the spy is keeping a count of all calls since it was created.
2. `const bar = d3.select(this).node();` is the same as saying `const bar = this;`, and the fact that `this` is the bar should be apparent from the `selectAll` above.

**Review by gordonwoodhull — CHANGES_REQUESTED** [review]
Fantastic work. Love to see thorough tests, and I finally tried it out via the new example and everything works flawlessly, as far as I can tell. The only thing I see you have not tested is newly added elements on other charts, but I think it's okay to skip that, given how tedious that is to test.

This is almost ready to merge, but I'd like to see the tests tightened up a little, and also the example clarified, as mentioned above.

File: src/base/base-mixin.js (hunk 6 above)
Comment (gordonwoodhull) [reply src/base/base-mixin.js]: I agree, there probably is no better way. I am not sure that checking `typeof d === 'object'` provides all that much protection, since objects are pretty common in JS.

Were you seeing errors if you didn't check this?

File: spec/bar-chart-spec.js (hunk 8 above)
Comment (gordonwoodhull) [reply spec/bar-chart-spec.js:1350]: Hmmm, thinking about this further, `toHaveBeenCalledWith` will check all calls ever. We 

want to see that the last call equals the datum, so I guess we need 

```js
expect(obj.method.calls.mostRecent().args[0]).toEqual(d)
```
to be precise.

Comment (gordonwoodhull) [reply spec/bar-chart-spec.js:1350]: I got it to work like this

```suggestion
            const clickHandlerSpy = jasmine.createSpy();
            chart.onClick = clickHandlerSpy;
            chart.selectAll('rect.bar').each(function (d) {
                this.dispatchEvent(event);
                expect(clickHandlerSpy).toHaveBeenCalledWith(d);
                clickHandlerSpy.calls.reset();
            });
```

Seems excessively complicated, but this clearly checks that the particular event caused the particular datum to get selected.

File: src/base/base-mixin.js (hunk 6 above)
Comment (gherka) [reply src/base/base-mixin.js]: I thought you were on holiday! :) Yes. on charts like boxplot where you have a group with bound data and a clickable circle element. If you press enter on the circle, the event triggers, then propagates up to the group. If we don't check for `d` type, the `onClick` function will try to run twice and the first call (from circle) will error out with `TypeError: filters[0] is undefined`.

File: src/charts/number-display.js [hunk 9]
```diff
@@ -121,36 +122,65 @@ export class NumberDisplay extends BaseMixin {
...
+                            let html = null;
+                            const num = chart.formatNumber()(interp(t));
+                            if (newValue === 0 && (chart._html.none !== '')) {
+                                html = chart._html.none;
+                            } else if (newValue === 1 && (chart._html.one !== '')) {
+                                html = chart._html.one;
+                            } else if (chart._html.some !== '') {
+                                html = chart._html.some;
+                            }
+                            node.innerHTML = html ? html.replace('%number', num) : num;
```
[the remark this answers is not in the document]
Comment (gherka) [reply src/charts/number-display.js:180]: I agree. That was my first thought to just set ```            if (this._ariaLiveRegion) {
                span
                    .attr('aria-live', 'polite')
                    .transition()
                    .duration(0);
            }```, but the transition still plays out - very quickly, but enough for the screen reader to read out the incorrect figure. There is a [question](https://stackoverflow.com/questions/14104423/how-to-make-true-zero-duration-instant-transitions-in-d3) on SO touching on this issue, but I thought it safer to just forego transition altogether at the expense of some code duplication.

File: web-src/examples/filtering.html [hunk 10]
```diff
@@ -64,12 +64,14 @@
...
     .dimension(yearDim)
     .group(spendPerYear)
+    .keyboardAccessible(true)
     .innerRadius(50)
     .controlsUseVisibility(true);
 
   spendHistChart
     .dimension(spendDim)
     .group(spendHist)
+    .svgDescription("Histogram")
```
[the remark this answers is not in the document]
Comment (gherka) [reply web-src/examples/filtering.html:74]: Not that I've seen yet. Probably possible with some keyboard modifiers like `Ctrl + Arrow Left / Right`. Will need to investigate properly. [This](https://bl.ocks.org/Andrew-Reid/1ba11865ee0d0585f0274a0976324358) looks quite promising.

File: spec/bar-chart-spec.js (hunk 8 above)
Comment (gherka) [reply spec/bar-chart-spec.js:1350]: Thanks, that's good to know. I'll apply your fix to all the other tests that check this functionality.

File: src/base/base-mixin.js (hunk 6 above)
Comment (gordonwoodhull) [reply src/base/base-mixin.js]: Thanks. Okay, let’s stick with the object check for now. Maybe we will figure something out (like another CSS class?) later.

File: src/charts/number-display.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (gordonwoodhull) [reply src/charts/number-display.js:180]: Please try using `dc.transition()` here, like we do most places. That should do the trick. I’m sure it was not skipped intentionally.

File: spec/bar-chart-spec.js (hunk 8 above)
Comment (gordonwoodhull) [reply spec/bar-chart-spec.js:1350]: Great, thanks!

File: web-src/examples/filtering.html [hunk 11]
```diff
@@ -64,12 +64,14 @@
 yearRingChart
```
[the remark this answers is not in the document]
Comment (gherka) [reply web-src/examples/filtering.html:65]: Done!

File: spec/bar-chart-spec.js (hunk 8 above)
Comment (gherka) [reply spec/bar-chart-spec.js]: Done!

File: src/charts/number-display.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (gherka) [reply src/charts/number-display.js]: I must've set the transition duration to zero incorrectly initially when testing. Works now with much cleaner code.

**Comment by gordonwoodhull** [comment]
Excellent work, thanks @gherka!

I squashed the commits and released this as 4.2.0.
